src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs

```diff
@@ -1116,22 +1138,77 @@ void GetPgoOffsetAndVersion(int offset, out int version, out int pgoDataOffset)
+                        continue;
+                    }
+
+                    int count = 0;
+                    int i = runtimeFunctionId;
+                    do
+                    {
+                        count++;
+                        i++;
+                    } while (i < isEntryPoint.Length && !isEntryPoint[i] && i < firstColdRuntimeFunction);
+
+                    method.ColdRuntimeFunctionId = dHotColdMap[runtimeFunctionId][0];
+                    method.RuntimeFunctionCount = count + coldSize;
+                    method.ColdRuntimeFunctionCount = coldSize;
+                }
+                else
+                {
+                    Debug.Assert(runtimeFunctionId < firstColdRuntimeFunction);
+                    if (runtimeFunctionId == -1)
+                        continue;
+
+                    int count = 0;
+                    int i = runtimeFunctionId;
+                    do
+                    {
+                        count++;
+                        i++;
+                    } while (i < isEntryPoint.Length && !isEntryPoint[i] && i < firstColdRuntimeFunction);
+                    method.RuntimeFunctionCount = count;
+                }
+            }
+        }
+
+        public void ValidateRuntimeFunctions(List<RuntimeFunction> runtimeFunctionList)
+        {
+            List<RuntimeFunction> runtimeFunctions = (List<RuntimeFunction>)runtimeFunctionList;
+            RuntimeFunction firstRuntimeFunction = runtimeFunctions[0];
+            BaseUnwindInfo firstUnwindInfo = firstRuntimeFunction.UnwindInfo;
+            var x64UnwindInfo = firstUnwindInfo as Amd64.UnwindInfo;
+            System.Collections.Generic.HashSet<uint> hashPersonalityRoutines = new System.Collections.Generic.HashSet<uint>();
+            if (x64UnwindInfo != null)
+            {
+                hashPersonalityRoutines.Add(x64UnwindInfo.PersonalityRoutineRVA);
+            }
 
-                int count = 0;
-                int i = runtimeFunctionId;
-                do
+            for (int i = 1; i < runtimeFunctions.Count; i++)
+            {
+                Debug.Assert(runtimeFunctions[i - 1].StartAddress.CompareTo(runtimeFunctions[i].StartAddress) < 0, "RuntimeFunctions are not sorted");
+                Debug.Assert(runtimeFunctions[i - 1].EndAddress <= runtimeFunctions[i].StartAddress, "RuntimeFunctions intervals overlap");
+
+                if (x64UnwindInfo != null && ((x64UnwindInfo.Flags & (int)ILCompiler.Reflection.ReadyToRun.Amd64.UnwindFlags.UNW_FLAG_CHAININFO) == 0))
                 {
-                    count++;
-                    i++;
-                } while (i < isEntryPoint.Length && !isEntryPoint[i]);
-                method.RuntimeFunctionCount = count;
+                    Amd64.UnwindInfo x64UnwindInfoCurr = (Amd64.UnwindInfo)runtimeFunctions[i].UnwindInfo;
+
+                    if ((x64UnwindInfoCurr.Flags & (int)ILCompiler.Reflection.ReadyToRun.Amd64.UnwindFlags.UNW_FLAG_CHAININFO) == 0)
```

Comment:
Looks like we already have access to the <code>UNW</code> constants, please consider removing their open-coded duplicates from <code>MethodGCInfoNode.cs</code>.

Reply:
It doesn't work this way. `ILCompiler.ReadyToRun.csproj` does not reference `ILCompiler.Reflection.ReadyToRun.csproj` and therefore we are unable to use the constants there.